presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -277,29 +278,53 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, PlanVariableAllocat
 
         private Multimap<VariableReferenceExpression, VariableReferenceExpression> extractCorrelatedSymbolsMapping(List<Expression> correlatedConjuncts)
         {
-            // TODO: handle coercions and non-direct column references
             ImmutableMultimap.Builder<VariableReferenceExpression, VariableReferenceExpression> mapping = ImmutableMultimap.builder();
             for (Expression conjunct : correlatedConjuncts) {
+                // TODO: Handle non-direct column references for ArithmeticBinaryExpression, ArithmeticUnaryExpression, NotExpression, etc
                 if (!(conjunct instanceof ComparisonExpression)) {
                     continue;
                 }
 
                 ComparisonExpression comparison = (ComparisonExpression) conjunct;
-                if (!(comparison.getLeft() instanceof SymbolReference
-                        && comparison.getRight() instanceof SymbolReference
-                        && comparison.getOperator().equals(EQUAL))) {
+                // handle coercions and non-direct column references
+                Expression leftExpression;
+                Expression rightExpression;
+                if (comparison.getLeft() instanceof Cast) {
```

Comment:
I think this whole code section would be better as an ExpressionVisitor so that it would handle nested expressions and such.

Reply:
yes makes sense, but instead of `ExpressionVisitor`, `AstVisitor` is used to traverse for expression like `Cast` or `SymbolReference`